src/main/java/org/opensearch/neuralsearch/processor/combination/ScoreCombiner.java

```diff
@@ -0,0 +1,161 @@
+        // iterate over results from each shard. Every CompoundTopDocs object has results from
+        // multiple sub queries, doc ids may repeat for each sub query results
+        return queryTopDocs.stream()
+            .map(compoundQueryTopDocs -> combineShardScores(scoreCombinationTechnique, compoundQueryTopDocs))
+            .collect(Collectors.toList());
+    }
+
+    private float combineShardScores(ScoreCombinationTechnique scoreCombinationTechnique, CompoundTopDocs compoundQueryTopDocs) {
+        if (Objects.isNull(compoundQueryTopDocs) || compoundQueryTopDocs.totalHits.value == 0) {
+            return ZERO_SCORE;
+        }
+        List<TopDocs> topDocsPerSubQuery = compoundQueryTopDocs.getCompoundTopDocs();
+        // - create map of normalized scores results returned from the single shard
+        Map<Integer, float[]> normalizedScoresPerDoc = getNormalizedScoresPerDocument(topDocsPerSubQuery);
+
+        // - create map of combined scores per doc id
+        Map<Integer, Float> combinedNormalizedScoresByDocId = combineScoresAndGetCombinedNormilizedScoresPerDocument(
+            normalizedScoresPerDoc,
+            scoreCombinationTechnique
+        );
+
+        // - sort documents by scores and take first "max number" of docs
+        // create a priority queue of doc ids that are sorted by their combined scores
+        PriorityQueue<Integer> scoreQueue = getPriorityQueueOfDocIds(normalizedScoresPerDoc, combinedNormalizedScoresByDocId);
+        // store max score to resulting list, call it now as priority queue will change after combining scores
+        float maxScore = combinedNormalizedScoresByDocId.get(scoreQueue.peek());
+
+        // - update query search results with normalized scores
+        updateQueryTopDocsWithCombinedScores(compoundQueryTopDocs, topDocsPerSubQuery, combinedNormalizedScoresByDocId, scoreQueue);
+        return maxScore;
+    }
+
+    private PriorityQueue<Integer> getPriorityQueueOfDocIds(
+        Map<Integer, float[]> normalizedScoresPerDoc,
+        Map<Integer, Float> combinedNormalizedScoresByDocId
+    ) {
+        PriorityQueue<Integer> pq = new PriorityQueue<>(
+            (a, b) -> Float.compare(combinedNormalizedScoresByDocId.get(b), combinedNormalizedScoresByDocId.get(a))
+        );
+        // we're merging docs with normalized and combined scores. we need to have only maxHits results
+        pq.addAll(normalizedScoresPerDoc.keySet());
+        return pq;
+    }
+
+    private ScoreDoc[] getCombinedScoreDocs(
+        final CompoundTopDocs compoundQueryTopDocs,
+        final Map<Integer, Float> combinedNormalizedScoresByDocId,
+        final PriorityQueue<Integer> scoreQueue,
+        final int maxHits
+    ) {
+        ScoreDoc[] finalScoreDocs = new ScoreDoc[maxHits];
+
+        int shardId = compoundQueryTopDocs.scoreDocs[0].shardIndex;
+        for (int j = 0; j < maxHits; j++) {
+            int docId = scoreQueue.poll();
```

Comment:
Is `maxHits` smaller than or equal to scoreQueue size?

Reply:
should be, but let me add isEmpty as second condition for loop